File: dotnet/DesignPatterns/BehavioralDesignPatterns/VisitorDesignPattern/VisitorDesignPattern.Example1/TaxCalculatorVisitor.cs
namespace VisitorDesignPattern.Example1;

/// <summary>
/// Concrete Visitor 1: Calculates the tax for each product type.
/// Electronics: 18% VAT
/// Food: 8% VAT (reduced rate)
/// Clothing: 12% VAT
/// </summary>
public class TaxCalculatorVisitor : IVisitor
{
    public decimal TotalTax { get; private set; }

    public void Visit(ElectronicsProduct product)
    {
        decimal tax = product.Price * 0.18m;
        TotalTax += tax;
        Console.WriteLine($"  [Tax] {product.Name}: ${product.Price:F2} x 18% VAT = ${tax:F2}");
    }

    public void Visit(FoodProduct product)
    {
        decimal taxRate = product.IsOrganic ? 0.05m : 0.08m;
        decimal tax = product.Price * taxRate;
        TotalTax += tax;
        string label = product.IsOrganic ? "5% (organic)" : "8%";
        Console.WriteLine($"  [Tax] {product.Name}: ${product.Price:F2} x {label} VAT = ${tax:F2}");
    }

    public void Visit(ClothingProduct product)
    {
        decimal tax = product.Price * 0.12m;
        TotalTax += tax;
        Console.WriteLine($"  [Tax] {product.Name}: ${product.Price:F2} x 12% VAT = ${tax:F2}");
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/VisitorDesignPattern/VisitorDesignPattern.Example1/src/TaxCalculatorVisitor.ts
import { IVisitor } from './IVisitor';
import { ElectronicsProduct } from './ElectronicsProduct';
import { FoodProduct } from './FoodProduct';
import { ClothingProduct } from './ClothingProduct';

export class TaxCalculatorVisitor implements IVisitor {
  private _totalTax: number = 0;

  get totalTax(): number {
    return this._totalTax;
  }

  visitElectronics(product: ElectronicsProduct): void {
    const tax = product.price * 0.18;
    this._totalTax += tax;
    console.log(`  [Tax] ${product.name}: $${product.price.toFixed(2)} x 18% VAT = $${tax.toFixed(2)}`);
  }

  visitFood(product: FoodProduct): void {
    const taxRate = product.isOrganic ? 0.05 : 0.08;
    const tax = product.price * taxRate;
    this._totalTax += tax;
    const label = product.isOrganic ? '5% (organic)' : '8%';
    console.log(`  [Tax] ${product.name}: $${product.price.toFixed(2)} x ${label} VAT = $${tax.toFixed(2)}`);
  }

  visitClothing(product: ClothingProduct): void {
    const tax = product.price * 0.12;
    this._totalTax += tax;
    console.log(`  [Tax] ${product.name}: $${product.price.toFixed(2)} x 12% VAT = $${tax.toFixed(2)}`);
  }
}
